utils/sb3/masked_buffer.py

```diff
@@ -0,0 +1,96 @@
+"""Module containing regularized PPO algorithm."""
+import logging
+from typing import Generator, Optional
+
+import numpy as np
+import torch
+from stable_baselines3.common.buffers import RolloutBuffer
+from stable_baselines3.common.type_aliases import RolloutBufferSamples
+
+logging.getLogger(__name__)
+
+
+class MaskedRolloutBuffer(RolloutBuffer):
+    """Custom SB3 RolloutBuffer class that filters out invalid samples."""
+
+    def compute_returns_and_advantage(self, last_values: torch.Tensor, dones: np.ndarray) -> None:
+        """GAE (General Advantage Estimation) to compute advantages and returns."""
+        # Convert to numpy
+        last_values = last_values.clone().cpu().numpy().flatten()
+
+        last_gae_lam = 0
+        for step in reversed(range(self.buffer_size)):
+            if step == self.buffer_size - 1:
+                # EDIT_1: Map NaNs to 1
+                dones = np.nan_to_num(dones, nan=1.0)
+
+                next_non_terminal = 1.0 - dones
+                next_values = last_values
+
+            else:
+                # EDIT_1: Map NaNs to 1
+                episode_starts = np.nan_to_num(self.episode_starts[step + 1], nan=1.0)
+
+                next_non_terminal = 1.0 - episode_starts
+                next_values = self.values[step + 1]
+
+            delta = (
+                np.nan_to_num(self.rewards[step], nan=0)  # EDIT_2: Set invalid rewards to zero
+                + np.nan_to_num(
+                    self.gamma * next_values * next_non_terminal, nan=0
+                )  # EDIT_3: Set invalid rewards to zero
+                - np.nan_to_num(self.values[step], nan=0)  # EDIT_4: Set invalid values to zero
```

Comment:
Leaving this as a comment to myself to review this carefully tomorrow to make sure this can't propagate values across episode boundaries